packages/plugin-ext/doc/how-to-add-new-plugin-namespace.md

```diff
@@ -90,23 +122,153 @@ function findPlugin(filePath: string): Plugin | undefined {
             fooBar : FooBar
         };
     }
+}
+```
+
+In the example above the API object creates a local object that will fulfill the API contract.
+The implementation details are hidden by the object and it could be a local implementation that only lives inside the plugin host but it could also be an implementation that uses the `RPCProtocol` to communicate with the main application to trigger changes, register functionality or retrieve information.
+
+### Implementing Main-Ext communication
+
+In this document, we will only highlight the individual parts needed to establish the communication between the main application and the external plugin host.
+For a more elaborate example of an API that communicates with the main application, please have a look at the definition of the [Theia Plugin API](https://github.com/eclipse-theia/theia/blob/master/doc/Plugin-API.md).
+
+First, we need to establish the communication on the RPC protocol by providing an implementation for our own side and generating a proxy for the opposite side.
+Proxies are identified using dedicated identifiers so we set them up first, together with the expected interfaces.
+`Ext` and `Main` interfaces contain the functions called over RCP and must start with `$`.
+Due to the asynchronous nature of the communication over RPC, the result should always be a `Promise` or `PromiseLike`.
+
+Example `common/foo-api-rpc.ts`:
+
+```typescript
+export interface FooMain {
+    $getFooImpl(): Promise<Foo>;
+}
+
+export interface FooExt {
+    // placeholder for callbacks for the main application to the extension
+}
+
+// Plugin host will obtain a proxy using these IDs, main application will register an implementation for it.
+export const PLUGIN_RPC_CONTEXT = {
+    FOO_MAIN: createProxyIdentifier<FooMain>('FooMain')
+};
+
+// Main application will obtain a proxy using these IDs, plugin host will register an implementation for it.
+export const MAIN_RPC_CONTEXT = {
+    FOO_EXT: createProxyIdentifier<FooExt>('FooExt')
+};
+```
+
+On the plugin host side we can register our implementation and retrieve the proxy as part of our `createAPIFactory` implementation:
+
+Example `foo-ext.ts`:
+
+```typescript
+export class FooExtImpl implements FooExt {
+    // Main application RCP counterpart
+    private proxy: FooMain;
+
+    constructor(rpc: RPCProtocol) {
+        rpc.set(MAIN_RPC_CONTEXT.FOO_EXT, this); // register ourselves
+        this.proxy = rpc.getProxy(PLUGIN_RPC_CONTEXT.FOO_MAIN); // retrieve proxy
+    }
+
+    getFooImpl(): Promise<Foo> {
+        return this.proxy.$getFooImpl();
+    }
+}
+```
+
+On the main side we need to implement the counterpart of the ExtPluginApiProvider, the `MainPluginApiProvider`, and expose it in a browser frontend module:
+
+Example `foo-main.ts`:
+
+```typescript
+@injectable()
+export class FooMainImpl implements FooMain {
+    protected proxy: FooExt;
+
+    init(rpc: RPCProtocol) {
+        this.proxy = rpc.getProxy(MAIN_RPC_CONTEXT.FOO_EXT);
```

Comment:
This example understandably doesn't provide an example of the main side of things calling back into the plugin-host. But it would be helpful to have a comment here, something along the lines of
```
# We would use this if we had a need to call back into the plugin-host/plugin.
this.proxy = rpc.getProxy(MAIN_RPC_CONTEXT.FOO_EXT);
```

Reply:
Good point, I added the comment you suggested.